ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+Such extention of the API that wraps namepsace has precedence, for example OpenShift apis provide a Project kind the wraps a k8s namespace.
+
+### Additional benefit 
+
+We can automate the bootstraping of namespaces with extra resources a konflux admin deems proper for their specific use case.
+
+We can create exension points that enable the community to use different backends of the implementation which will include the ability to scale to multiple cluster with a single instance of Konflux.
+
+
+## Konflux Workspaces Upstream Design Proposal
+
+### Purpose
+
+This a design proposal to represent the Konflux workspace concept long term. It aims to show how the Konflux upstream project should implement a workspace abstraction that will allow the community to select their own authentication provider as well as their own namespace provisioning backend. Three namespace provisioning backends will be presented as reference implementation but it should not be limited to those as the community will be able to extend or modify to better suit their use case.
+
+### Glossary
+
+Workspace: A collection of resources in a namespace that are required for Konflux secured pipelines to execute.
+
+NSaaS: A service that provisions, bootstraps, and manages k8s namespaces.
+
+NSaaS backend: A specific implementation of a workspace controller.
+
+### Background
+
+Konflux Workspaces is an abstraction for a Kubernetes namespace containing a set of predefined resources that a Konflux user needs for their secured pipeline lifecycle operation. The workspace concept was introduced by AppStudio when KCP was considering each workspace being its own virtual Kubernetes cluster, and it was later implemented via KubeSaw as a NSaaS technology used by [Developer Sandbox](https://developers.redhat.com/developer-sandbox). Today, the concept continues to be the same: each user in Konflux needs at least one namespace in some cluster where their pipeline jobs can execute.
+
+Directly managing a namespace for a user has limitations imposed by Kubernetes.  First, Kubernetes lacks the granularity to allow a user to see only the namespaces it has access to.  If we want to only provide a list of namespace a user has access to, we will need another mechanism.  Second, we also will need to support provisioning of a namespace, which involves the creation of the namespace and populating it with the resources necessary for Konflux including but not limited to RoleBindings and Quotas.  Without any kind of secondary mechanism for this process, a user will need to understand how to do this process manually or delegate to an admin.  Finally, we will want to have support for setting specific limmits and restrictions on users, such as how many workspaces a user can create and own.  Collecting these metrics and determining violations will require some mechanism for enforcing these policies, which Kubernetes does not give us.
```

Comment:
1. Do you see any reasons why we can't achieve it by using namespace directly + keyverno/argocd, or as mentioned above in openshift using projects?

2. Why do you think it's an issue for Konflux to solve?  for example, when I install (kubevirt)[https://kubevirt.io/], it doesn't create a namespace for me where I can create VMs in. Same goes for Tekton, I doesn't create a namespace for me where I can run piplines. TBH, I'm not aware of any extension that provides this functionality, why Konflux is special?

Reply:
> 1. Do you see any reasons why we can't achieve it by using namespace directly + keyverno/argocd, or as mentioned above in openshift using projects?

There are many solution as you mentioned that involved using existing controllers that reacts to user input. I just think that this approach provides the simples solution, for the admin and for the end user. End user is particularly benefiting from this as it will only need to learn how to manage a single workspace CR. Documentation will be straight forward as well. 

> 2. Why do you think it's an issue for Konflux to solve?  for example, when I install (kubevirt)[https://kubevirt.io/], it doesn't create a namespace for me where I can create VMs in. Same goes for Tekton, I doesn't create a namespace for me where I can run piplines. TBH, I'm not aware of any extension that provides this functionality, why Konflux is special?

From the beginning the way I think about konflux is technology put together to implement a secured CI/CD service, not as a product. In services the on-boarding and off-boarding of users are a cornerstone piece of the puzzle, providing consistency that can enable other services to integrate with it. One advantage of us providing this solution in the community is that we will benefit on Red Hat deployments from the community ideas, and the community will benefit from our Red Hat reference implementation.